### PR TITLE
Move memcached-operator monitoring practices to a separate operator

### PR DESCRIPTION
**Pull request by assafad**
Signed-off-by: assaf-admi <aadmi@redhat.com>

**Description of the change:**
Moving everything related to monitoring in the `memcached-operator` sample to a new separate sample `monitoring/memcached-operator`, which will include monitoring practices by default (metrics, alerts, etc). 
The suggested structure for the samples would be:
```
testdata
|_ go
    |_ v3
        |_ memcached-operator
        |_ monitoring
            |_ memcached-operator
    |_ v4-alpha
        |_ memcached-operator
        |_ monitoring
            |_ memcached-operator
        
```

**Motivation for the change:**
https://github.com/operator-framework/operator-sdk/issues/6142

**Checklist**

If the pull request includes user-facing changes, extra documentation is required:
- [ ] Add a new changelog fragment in `changelog/fragments` (see [`changelog/fragments/00-template.yaml`](https://github.com/operator-framework/operator-sdk/tree/master/changelog/fragments/00-template.yaml))
- [ ] Add or update relevant sections of the docs website in [`website/content/en/docs`](https://github.com/operator-framework/operator-sdk/tree/master/website/content/en/docs)

### CODE REVIEW

**Comment by assafad** [comment]
cc @sradco, @avlitman, @machadovilaca, @everettraven, @varshaprasad96

**Comment by varshaprasad96** [comment]
@assafad Took a first glance at the PR and it does look good. Since the monitoring scaffoldings were a part of go/v3 previously and hence were tested in the cluster, can we add this project also to Makefile: https://github.com/operator-framework/operator-sdk/blob/a163a1b3f25c6322c806e96adccfd8936a634511/Makefile#L190-L206, to run initial scaffolding tests and be sure that nothing breaks.

**Comment by varshaprasad96** [comment]
Looks like the go e2e tests copy the folder from testdata/go/v3/memcached: https://github.com/operator-framework/operator-sdk/blob/a163a1b3f25c6322c806e96adccfd8936a634511/test/e2e/go/suite_test.go#L59 and run the tests. Actually its fine, we can skip the e2e tests for now. We need to make the go tests configurable to accept any project path which is currently hardcoded there. That means we need to spend additional CI cycles to ensure that our scaffolded samples are tested. We can do this in a follow up.

**Review by varshaprasad96 — APPROVED** [review]
Rest if the changes of creating v3 and v4-alpha monitoring  + reverting both of them to have no additions related to metrics looks good to me!
/lgtm

cc: @everettraven @jmrodri please feel free to take a look

**Comment by avlitman** [comment]
@assafad can you share the steps for verifying this pr?
Not sure why the sanity failed, other than that lgtm.

**Comment by assafad** [comment]
> @assafad can you share the steps for verifying this pr?
> Not sure why the sanity failed, other than that lgtm.

Hi @avlitman, for testing the PR I generated the operators with ‘make generate’, deployed each of them on a kind cluster, and verified the operator works with no errors. In addition, I ran ‘make test-e2e’ for each operator and it didn’t failed.
I will have to take a deeper look and see why the sanity failed.

**Comment by avlitman** [comment]
> > @assafad can you share the steps for verifying this pr?
> > Not sure why the sanity failed, other than that lgtm.
> 
> Hi @avlitman, for testing the PR I generated the operators with ‘make generate’, deployed each of them on a kind cluster, and verified the operators works with no errors. In addition, I ran ‘make test-e2e’ for each operator and it didn’t fail. I will have to take a deeper look and see why the sanity failed.

Great, Thank you for sharing the steps!

**Comment by assafad** [comment]
@varshaprasad96, @everettraven I ran `make test-sanity` locally and it modified a few YAML files. Should I push these changes in order to pass the sanity check that failed?

**Comment by everettraven** [comment]
> I ran `make test-sanity` locally and it modified a few YAML files in `v4-alpha/memcached operator` (e.g. `v4-alpha/memcached-operator/bundle/manifests/memcached-operator.clusterserviceversion.yaml`). After that, I ran the operator's `make test-e2e` and it passed locally, so I believe it didn't break anything.
> 
> Should I push these changes in order to pass the sanity check that failed?

@assafad Yes, those changes should be safe to push up.

**Comment by assafad** [comment]
> > I ran `make test-sanity` locally and it modified a few YAML files in `v4-alpha/memcached operator` (e.g. `v4-alpha/memcached-operator/bundle/manifests/memcached-operator.clusterserviceversion.yaml`). After that, I ran the operator's `make test-e2e` and it passed locally, so I believe it didn't break anything.
> > Should I push these changes in order to pass the sanity check that failed?
> 
> @assafad Yes, those changes should be safe to push up.

Done, let's see if it passes.

**Comment by assafad** [comment]
@everettraven, @varshaprasad96 All tests have passed, anything else we're missing in order to merge it?

File: testdata/go/v4-alpha/monitoring/memcached-operator/test/e2e/e2e_test.go [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+		It("MemcachedDeploymentSizeUndesiredCountTotal should be increased when scaling the Memcached deployment", func() {
+			initialMetricValue := getMetricValue(memcachedDeploymentSizeUndesiredCountTotalName)
+
+			numberOfScales := 5
+			By(fmt.Sprintf("scaling memcached-samle deployment %d times", numberOfScales))
+			scaleMemcachedSampleDeployment(numberOfScales)
+
+			By(fmt.Sprintf("validating MemcachedDeploymentSizeUndesiredCountTotal has increased by %d", numberOfScales))
+			finalMetricValue := getMetricValue(memcachedDeploymentSizeUndesiredCountTotalName)
+			Expect(finalMetricValue).To(Equal(initialMetricValue + numberOfScales))
```
Comment (everettraven) [inline testdata/go/v4-alpha/monitoring/memcached-operator/test/e2e/e2e_test.go:240]: Ditto

**Review by everettraven — COMMENTED** [review]
@assafad These changes look great! Thanks so much for the contribution!

I only have one concern that I would like to get some thoughts on prior to merging:

File: testdata/go/v3/monitoring/memcached-operator/test/e2e/e2e_test.go [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+		It("MemcachedDeploymentSizeUndesiredCountTotal should be increased when scaling the Memcached deployment", func() {
+			initialMetricValue := getMetricValue(memcachedDeploymentSizeUndesiredCountTotalName)
+
+			numberOfScales := 5
+			By(fmt.Sprintf("scaling memcached-samle deployment %d times", numberOfScales))
+			scaleMemcachedSampleDeployment(numberOfScales)
+
+			By(fmt.Sprintf("validating MemcachedDeploymentSizeUndesiredCountTotal has increased by %d", numberOfScales))
+			finalMetricValue := getMetricValue(memcachedDeploymentSizeUndesiredCountTotalName)
+			Expect(finalMetricValue).To(Equal(initialMetricValue + numberOfScales))
```
[the remark this answers is not in the document]
Comment (assafad) [reply testdata/go/v3/monitoring/memcached-operator/test/e2e/e2e_test.go:240]: I'm reviewing some ways to fix this test's flakiness. Planned to include it https://github.com/operator-framework/operator-sdk/pull/6135. @everettraven Does it make sense?

[the remark this answers is not in the document]
Comment (assafad) [reply testdata/go/v3/monitoring/memcached-operator/test/e2e/e2e_test.go:240]: @everettraven For example, one possible solution would be to wait more time after scaling the deployment https://github.com/operator-framework/operator-sdk/blob/master/testdata/go/v4-alpha/memcached-operator/test/e2e/e2e_test.go#L385

[the remark this answers is not in the document]
Comment (everettraven) [reply testdata/go/v3/monitoring/memcached-operator/test/e2e/e2e_test.go:240]: >I'm reviewing some ways to fix this test's flakiness. Planned to include it in https://github.com/operator-framework/operator-sdk/pull/6135. I think we should merge this relocating PR if we can, so we can move ahead with other patches and fix that test in a more precise way. @everettraven Does it make sense?

@assafad that sounds good to me. Just wanted to make sure it was on the radar :)

**Review by everettraven — APPROVED** [review]
/lgtm

**Pull request merged**